protocols/index.md

```diff
@@ -0,0 +1,33 @@
+# CloudState Documentation and Specification format rules
+
+The CloudState Protocol Specification MUST use [RFC-2119](https://www.ietf.org/rfc/rfc2119.txt) definitions.
+
+Use the following formatting and Markdown syntax to define documentation and specification of the CloudState Protocol.
+The template/example below is proven to work well with the `protoc-gen-doc` plugin, so please stick to it 100%.
+
+```
+//The *EntityDiscovery* Service is the entrypoint for CloudState Proxies
+//
+//---
+//
+//`Specification`:
+//
+//  - A `User Function` MUST implement EntityDiscovery
+//  - `discover`
+//    - *MUST* be invoked with a *valid* `ProxyInfo`
+//    - *MUST* be *idempotent*
+//    - *MAY* be called any number of times by its CloudState Proxy
+//    - *MUST* yield a failure if the provided `ProxyInfo` is not deemed compatible with the User Function
+//  - `reportError`
+//    - *MUST* only be invoked to inform a User Function about violations of the CloudState Specification
+//    - *MUST* only receive failures not originating from invoking `reportError`, since that would cause a crash-loop.
+//    - implementations are *RECOMMENDED* to log the received errors, for debugging purposes.
+service EntityDiscovery {
+
+    // Lets a CloudState Proxy discover which `service` and `entities` a User Function implements.
+    rpc discover(ProxyInfo) returns (EntitySpec) {}
+
+    // Lets a CloudState Proxy inform a User Function about specification violations.
+    rpc reportError(UserFunctionError) returns (google.protobuf.Empty) {}
```

Comment:
In what state would a user function be after such an error? If the implementation would not be compliant with `violations of the CloudState Specification` how should it act after this.

Also, such an error would not have any context at all. So does "inform" mean, that the proxy would passivate this entity? See also the discussion in go-support  https://github.com/cloudstateio/go-support/pull/5/files#r337442535

Reply:
That's a great point, @marcellanz. We should spec the intended sequence of actions after `reportError` has been called. It would seem like it is a fatal error.

Reply:
@viktorklang I have the impression, the proxy should passivate the user function, and re-initialise the entity. The user function might behave the same the second time and the proxy might re-try a few times. I don't know what's the proxies strategy here. Does the proxy force-restart user functions completely?

`reportError` has a symmetry with a `protocol.Failure` here, where the user function can decide how to proceed. It can report the failure and then close the stream and go on. It could even panic/os.exit(70) itself and get be re-started.

It seems on both occasions, on `reportError` and `protocol.Failure`, the other side only can log what it got reported, but the error detecting sending party knows best how to proceed having the error and its context at hand.

For the Go and its panic semantics I noted and got back to this part of the spec.
```
// there are two ways to do it
// a) report and close the stream and keep others running
// b) report and panic and therefore crash the program
// how can we decide that the panic keeps the user function in
// a consistent state. this one occasion could be perfectly ok
// to crash, but thousands of others kept running. why get them all down?
// so then, there is the presumption that a panic is truly exceptional
// and we can't be sure about anything to be safe after a panic.
// The proxy is well prepared for this, it is able to re-establish 
// and also isolate the erroneous entity type from others after failures.
```

It would be interesting to know what is the intended flow of events in such a scenario. The simpler the better perhaps. As the user functions gets its state from the proxy, the proxy also should know best what to do.

(note: `reportError` is not called `reportFailure`)